CONTRIBUTING.rst

```diff
@@ -0,0 +1,461 @@
+various possible errors.  None of these callbacks are expected to return a
+value.  Every use case should define an associated UI as an abstract base
+class.
+
+The call in the request handler on the UI side looks like:
+
+.. code:: python
+
+    usecase = usecase_factory.create_disable_permission_usecase(user, self)
+    usecase.disable_permission(name)
+
+This does not return a value.  Instead, the handler chains a call to the
+use case, which then chains a call to one of the callbacks depending on
+the result of the use case.  The callback in the UI is then responsible
+for returning the results to the user.  All of these calls return `None`.
+(Note that this requires a UI framework, like Tornado, that can return
+results to a user without relying on method return values.)
+
+If the use case needs to pass more complex data back to the UI, such as
+lists of objects, that data is represented by a data transfer object
+defined in `grouper.entities`.  These are mypy `NamedTuples` so that we
+have type checking of their attributes.  Always use data transfer objects
+like this, rather than objects that are artifacts of the underlying
+repository or have any behavior attached (in other words, don't return
+SQLAlchemy models).
+
+For paginated lists of objects, define an `Enum` to represent the possible
+sort keys and then use the generic types defined in
+`grouper.entities.pagination`.  The UI passes in a `Pagination` object as
+one of the arguments to the use case, and the use case passes a
+`PaginatedList` object as an argument to the success callback.
+
+Use Case to Service
+-------------------
+
+Use cases themselves make decisions but do not change or query data
+stores.  The mechanics of the requested operation is done via services
+and, underneath the services, repositories.
+
+The UI to use case call has a natural dependency direction: the UI depends
+on the use case and implements the use case's UI interface.  The use case
+to service call would naturally produce a dependency on the service from
+the use case.  But the goal of this architecture is to isolate the use
+case from the surrounding layers and ensure it doesn't depend on any given
+implementation.  The natural dependency direction is therefore inverted
+via dependency injection: when a use case is created, the services it
+needs are provided as arguments to its constructor.
+
+Every service a use case needs has a corresponding interface defined in
+`grouper.usecases.interfaces`.  Often there is only one implementation of
+that interface.  By convention, the single implementation is called
+`FooService` and the interface `FooInterface`.
+
+Use cases should make all *decisions*, including authorization, policy,
+and enforcing invariants such as "you cannot disable a system permission."
+Services should do all *work*, such as changing rows in databases,
+gathering data and converting it to data transfer objects, and so forth.
```

Comment:
It says services covert data to data transfer objects here, but changes its mind in the next section. Maybe re-phrase to indicate this is the interface they provide, but the actual work might be delegated?